benchmark/main.cpp

```diff
@@ -56,979 +50,1198 @@
+                                                        NULL, 0, buf, buf, tmp,
+                                                        16));
     }
-#endif
-#if defined(MBEDTLS_CMAC_C)
-    if( todo->aes_cmac )
-    {
-        unsigned char output[16];
-        const mbedtls_cipher_info_t *cipher_info;
-        mbedtls_cipher_type_t cipher_type;
-        int keysize;
-
-        cipher_type = MBEDTLS_CIPHER_AES_128_ECB;
-        for( keysize = 128; keysize <= 256; keysize += 64 )
-        {
-            mbedtls_snprintf( title, sizeof( title ), "AES-CMAC-%d", keysize );
-
-            memset( buf, 0, sizeof( buf ) );
-            memset( tmp, 0, sizeof( tmp ) );
-
-            cipher_info = mbedtls_cipher_info_from_type( cipher_type );
-
-            TIME_AND_TSC( title,
-                          mbedtls_cipher_cmac( cipher_info, tmp, keysize,
-                                               buf, BUFSIZE, output ) );
-            cipher_type = (mbedtls_cipher_type_t)( cipher_type + 1 );
+
+exit:
+    mbedtls_ccm_free(&ccm);
+#endif /* MBEDTLS_AES_C && MBEDTLS_CCM_C */
+
+    return ret;
+}
+
+MBED_NOINLINE static int benchmark_aes_cmac() {
+    int ret = 0;
+#if defined(MBEDTLS_AES_C) && defined(MBEDTLS_CMAC_C)
+    unsigned char output[16];
+    const mbedtls_cipher_info_t *cipher_info;
+    mbedtls_cipher_type_t cipher_type;
+    int keysize;
+
+    cipher_type = MBEDTLS_CIPHER_AES_128_ECB;
+    for (keysize = 128; keysize <= 256; keysize += 64) {
+        ret = mbedtls_snprintf(title, sizeof(title), "AES-CMAC-%d", keysize);
+        if (ret < 0 || static_cast<size_t>(ret) >= sizeof(title)) {
+            mbedtls_printf("Failed to compose title string using "
+                           "mbedtls_snprintf(): %d\n", ret);
+            goto exit;
         }
 
-        memset( buf, 0, sizeof( buf ) );
-        memset( tmp, 0, sizeof( tmp ) );
-        TIME_AND_TSC( "AES-CMAC-PRF-128",
-                      mbedtls_aes_cmac_prf_128( tmp, 16, buf, BUFSIZE,
-                                                output ) );
+        memset(buf, 0, sizeof(buf));
+        memset(tmp, 0, sizeof(tmp));
+
+        cipher_info = mbedtls_cipher_info_from_type(cipher_type);
+        if (cipher_info == NULL) {
+            mbedtls_printf("mbedtls_cipher_info_from_type() returned NULL\n");
+            goto exit;
+        }
+
+        BENCHMARK_FUNC_CALL(title,
+                            mbedtls_cipher_cmac(cipher_info, tmp, keysize,
+                                                buf, BUFSIZE, output));
+        cipher_type = (mbedtls_cipher_type_t)(cipher_type + 1);
     }
-#endif /* MBEDTLS_CMAC_C */
-#endif
 
+    memset(buf, 0, sizeof(buf));
+    memset(tmp, 0, sizeof(tmp));
+
+    BENCHMARK_FUNC_CALL("AES-CMAC-PRF-128",
+                        mbedtls_aes_cmac_prf_128(tmp, 16, buf, BUFSIZE,
+                        output));
+
+exit:
+#endif /* MBEDTLS_AES_C && MBEDTLS_CMAC_C */
+
+    return ret;
+}
+
+MBED_NOINLINE static int benchmark_camellia() {
```

Comment:
why not surround the whole function with `#if defined(MBEDTLS_CAMELLIA_C) && defined(MBEDTLS_CIPHER_MODE_CBC)`?